File: app/src/main/java/com/devusercode/upchat/utils/UserUtils.java
package com.devusercode.upchat.utils;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.devusercode.upchat.models.User;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

public class UserUtils {
    private static final String STAG = "UserUtils";
    private static final String REF = "users";

    public static class Result {
        public User user;
        public Error error;

        public Result(User user, Error error) {
            this.user = user;
            this.error = error;
        }

        public Result(User user) {
            this.user = user;
            this.error = null;
        }

        public User getUser() {
            return user;
        }

        public Error getError() {
            return error;
        }

        public boolean isSuccessful() {
            return error == null && user != null;
        }
    }

    public interface UserCallback {
        void onUserReceived(User user);

        void onUserNotFound();

        void onUserError(String errorMessage);
    }

    public static void getUserByUid(String uid, final UserCallback callback) {
        DatabaseReference usersRef = FirebaseDatabase.getInstance().getReference(REF).child(uid);

        usersRef.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if (dataSnapshot.exists()) {
                    User user = dataSnapshot.getValue(User.class);

                    if (user != null) {
                        callback.onUserReceived(user);
                    } else {
                        callback.onUserError("Unknown error: error while retrieving the user!");
                    }
                } else {
                    callback.onUserNotFound();
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
                callback.onUserError(databaseError.getMessage());
            }
        });
    }

    public static void getUserByUid(String uid, @NonNull Consumer<Result> onFinish) {
        DatabaseReference usersRef = FirebaseDatabase.getInstance().getReference(REF).child(uid);

        usersRef.get().addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                DataSnapshot dataSnapshot = task.getResult();

                if (dataSnapshot != null && dataSnapshot.exists()) {
                    User user = dataSnapshot.getValue(User.class);

                    if (user != null) {
                        onFinish.accept(new Result(user));
                    } else {
                        Error error = new Error("Unknown error: error while retrieving the user!");
                        Log.e(STAG, error.getMessage());

                        onFinish.accept(new Result(null, error));
                    }
                } else {
                    Error error = new Error("UserNotFound");
                    Log.e(STAG, error.getMessage());

                    onFinish.accept(new Result(null, error));
                }
            } else {
                Error error = new Error(task.getException().getMessage());
                Log.e(STAG, error.getMessage());

                onFinish.accept(new Result(null, error));
            }
        });
    }

    public static void update(String field, Object value) {
        String uid = FirebaseAuth.getInstance().getCurrentUser().getUid();

        DatabaseReference ref = FirebaseDatabase.getInstance().getReference()
                .child("users")
                .child(uid);

        ref.child(field).setValue(value).addOnFailureListener(error -> Log.e(STAG, error.getMessage()));
    }
}

File: app/src/main/java/com/devusercode/upchat/utils/UserUtils.kt
package com.devusercode.upchat.utils

import android.util.Log
import com.devusercode.upchat.models.User
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.util.function.Consumer

enum class ErrorCodes(val message: String) {
    USER_NOT_FOUND("User not found"),
    SUCCESS("Success"),
    UNKNOWN_ERROR("Unknown error"),
    INIT("Initialization error"),
    CANCELLED("Operation cancelled"),
    NO_PROP_TO_SERIALIZE("No properties to serialize found on class"),
    NO_PROP_TO_SERIALIZE_FOR_FIELD("No properties to serialize for field"),
}

class UserUtils {
    companion object {
        private const val TAG = "UserUtils"
        private const val REF = "users"

        // Result class to encapsulate user-fetching results
        class Result(
            val user: User?,
            val error: Error?,
            val code: ErrorCodes
        ) {
            constructor(user: User, code: ErrorCodes) : this(user, null, code)
            constructor(error: Error, code: ErrorCodes) : this(null, error, code)

            val isSuccessful: Boolean
                get() = user != null && error == null && code == ErrorCodes.SUCCESS
        }

        /**
         * Fetch user by UID using callback.
         * @param uid User's unique ID.
         * @param onFinish Callback to handle the result.
         */
        fun getUserByUid(uid: String, onFinish: Consumer<Result>?) {
            val usersRef = FirebaseDatabase.getInstance().getReference(REF).child(uid)

            CoroutineScope(Dispatchers.IO).launch {
                val result = fetchUser(uid)
                withContext(Dispatchers.Main) {
                    onFinish?.accept(result)
                }
            }
        }

        /**
         * Fetch user by UID as a suspend function.
         * @param uid User's unique ID.
         * @return Result containing user or error.
         */
        suspend fun getUserByUidAsync(uid: String): Result {
            return fetchUser(uid)
        }

        /**
         * Update a specific field for the current user.
         * @param field The field to update.
         * @param value The new value.
         */
        fun update(field: String?, value: Any?) {
            val currentUser = FirebaseAuth.getInstance().currentUser
            if (currentUser == null || field == null || value == null) {
                Log.e(TAG, "Invalid input or user is not authenticated")
                return
            }

            val uid = currentUser.uid
            val ref = FirebaseDatabase.getInstance().reference.child(REF).child(uid)

            ref.child(field).setValue(value)
                .addOnFailureListener { error -> Log.e(TAG, "Update failed: ${error.message}") }
        }

        // Internal helper function to fetch user
        private suspend fun fetchUser(uid: String): Result {
            val usersRef = FirebaseDatabase.getInstance().getReference(REF).child(uid)

            return try {
                val dataSnapshot = usersRef.get().await()
                if (dataSnapshot.exists()) {
                    val user = dataSnapshot.getValue(User::class.java)
                    if (user != null) {
                        Result(user, ErrorCodes.SUCCESS)
                    } else {
                        Result(Error("Failed to parse user data"), ErrorCodes.UNKNOWN_ERROR)
                    }
                } else {
                    Result(Error("User not found: $uid"), ErrorCodes.USER_NOT_FOUND)
                }
            } catch (exception: Exception) {
                Result(Error(exception.message ?: "Unknown error"), ErrorCodes.CANCELLED)
            }
        }
    }
}
